src/main/java/com/klodnicki/Bike/v2/service/BikeService.java

```diff
@@ -1,7 +1,126 @@
+import com.klodnicki.Bike.v2.repository.BikeRepository;
+import org.modelmapper.ModelMapper;
 import org.springframework.stereotype.Service;
 
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+
 @Service
-public class BikeService {
+//public class BikeService implements GenericService<BikeForAdminResponseDTO, Bike> {
+//zamienilismy BikeService dependency wstrzykiwane w controllerze na GenericBikeService
+//Żeby móc podstawić różne implementacje BikeService, nie tylko ten jeden konkretny BikeService. Jak zrobisz klasę
+//        BetterBikeService możesz bez żadnych zmian w kontrolerze z niego korzystać od razu
+public class BikeService implements GenericBikeService {
+
+    private final BikeRepository bikeRepository;
+
+    ModelMapper modelMapper = new ModelMapper();
+
+    public BikeService(BikeRepository bikeRepository) {
+        this.bikeRepository = bikeRepository;
+    }
+
+    @Override
+    public BikeForAdminResponseDTO add(BikeRequestDTO bikeRequestDTO) {
+        //using ModelMapper to convert(map) bikeRequestDTO into bike. I need Bike to be able to save it in repo
+        Bike bike = modelMapper.map(bikeRequestDTO, Bike.class);
+        Bike savedBike = bikeRepository.save(bike);
+
+        return convertBikeIntoBikeForAdminResponseDTO(savedBike); // gdyby tutaj brał bike (a nie savedBike) to by klasa
+        //abstrakcyjna nie musiała mieć adnotacji @MappedSuperclass
+    }
+
+    @Override
+    public BikeForAdminResponseDTO findById(Long id) {
+        Bike savedBike = bikeRepository.findById(id).orElseThrow(IllegalArgumentException::new); //TODO: add custom exception
+
+        return convertBikeIntoBikeForAdminResponseDTO(savedBike);
+    }
+
+    @Override
+    public List<BikeForAdminResponseDTO> findAll() {
+        List<BikeForAdminResponseDTO> bikeListDto = new ArrayList<>();
+        Iterable<Bike> bikeList = bikeRepository.findAll();
+
+        for (Bike bike : bikeList) {
+
+            Long id = bike.getId();
+            String serialNumber = bike.getSerialNumber();
+            boolean isRented = bike.isRented();
+            BikeType type = bike.getBikeType();
+            LocalDateTime rentalStartTime = bike.getRentalStartTime();
+            LocalDateTime rentalEndTime = bike.getRentalEndTime();
+            double toBePaid = bike.getAmountToBePaid();
+            GpsCoordinates gps = bike.getGpsCoordinates();
+
+            BikeForAdminResponseDTO bikeDto = new BikeForAdminResponseDTO(id, serialNumber, isRented, type, rentalStartTime,
+                    rentalEndTime, toBePaid, gps);
+            bikeListDto.add(bikeDto);
+        }
+
+        return bikeListDto;
+    }
+
+    @Override
+    public void deleteById(Long id) {
+        bikeRepository.deleteById(id);
+    }
+
+    @Override
+    public BikeForAdminResponseDTO update(Long id, BikeRequestDTO updatedBikeRequestDTO) {
+
+        Bike bike = bikeRepository.findById(id).orElseThrow(IllegalAccessError::new);
```

Comment:
Usage of IllegalAccessError is wrong here. You can check what creator of this Error meant here by getting into this class and checking description above which says:

_"Thrown if an application attempts to access or modify a field, or to call a method that it does not have access to.
Normally, this error is caught by the compiler; this error can only occur at run time if the definition of a class has incompatibly changed."_

Also a lesson - Error is **NOT** the same as Exception! This is crucial to know as this is often a question on job interview!

Reply:
this was made by mistake. I wanted to have here IllegalArgumentException until custom exception is not ready